Edit profile

jacek15157 opened the pull request:
logic for button "Potwierdz zmiany" and "Usun konto" in EditProfile.
"Potwierdz zmiany" can create new user (registration) or save changes if profile already exists.
"Usuń konto" is a soft delete for existing account.

kobrynsky committed:
resolved namespaces conflict

kobrynsky committed:
added phoneNumber to user model

kobrynsky committed:
Changed name of dto. Added required props to dto.

kobrynsky committed:
implementation of regsiter/edit profile button

jacek15157 committed:
soft delete user account

jacek15157 committed:
Change name of folders, service and controller from 'Login' to 'User'

jacek15157 committed:
fixed name of controller's prop

jskurzynska commented on src/Joblify.Core.Api/User/UserController.cs:
```diff
@@ -0,0 +1,55 @@
+
+        [HttpPost("saveProfile")]
+        [ValidateModel]
+        public async Task<IActionResult> SaveProfile([FromBody] EditProfileDto dto)
```
why is it "EditProfileDto" if the method is SaveProfile? DTO should be used only for the specific action

jskurzynska commented on src/Joblify.Core.Api/User/UserController.cs:
```diff
@@ -0,0 +1,55 @@
+        [HttpDelete("deleteProfile")]
+        public async Task<IActionResult> DeleteUser(string email)
+        {
+            var userFromRepo = await _userService.GetUser(email);
```
maybe "existedUser"?

jskurzynska commented on src/Joblify.Core.Api/User/UserController.cs:
```diff
@@ -0,0 +1,55 @@
+        [ValidateModel]
+        public async Task<IActionResult> SaveProfile([FromBody] EditProfileDto dto)
+        {
+            var result = await _userService.SaveProfile(dto);
```
I'm not even sure if it's SaveProfile or it should be "AddUser"?

jskurzynska commented on src/Joblify.Core.Api/User/UserController.cs:
```diff
@@ -0,0 +1,55 @@
+        }
+
+        [HttpDelete("deleteProfile")]
+        public async Task<IActionResult> DeleteUser(string email)
```
I know that we don't have time so for now it's ok, but if we want to have REST API, we should use only Get, Post, Put, Delete methods specific for User in that case. So, if we want to check if email is available there should be controller specific for that. In current situation we would have two GET methods if we wanted to return User info.

EnjoyTheNoise replied:
Why? Isn't checking email (which is in our system a way of checking if _user_ exists) perfectly fine to be here? I don't think I understand fully last sentence tho, can you please elaborate?

Edit: to clarify, I understand that having two gets with just email in query would end in getting nice big question mark from framework, but I think if we want to get specific user, Id would be more useable - question is, whether we want to create whole controller existing solely for checking if user exists (outside of UserController context), or to bend REST rules a bit, and create specific get endpoint within UserController 🤔

jacek15157 replied:
Yesterday, I fixed some parts of code below. Please tell if am I supposed to do something more with this.

jacek15157 committed:
Fixed rape on REST rules.

EnjoyTheNoise commented on src/Joblify.Core/User/UserService.cs:
```diff
@@ -0,0 +1,71 @@
+            var userEntity = _unitOfWork.UserRepository.Entities.SingleOrDefault(u => u.Email == userDto.Email);
+            if (userEntity == null)
+            {
+                throw new Exception("User does not exist");
```
why throw?

jacek15157 replied:
Agree. In fact I handle in controler posibility of getting null from repository. So it is redundancy here. However if I need to do this here, should I use try catch block with NullReferenceException?

EnjoyTheNoise replied:
First of all you should think if you really need to throw something - here we have just null being returned, so it's perfect candidate for either 400 or 404 response.
Also, to address your question - yes, if you want to react to specific exceptions, you have to catch them and handle application flow in catch or finally (or middleware designated to handle server-side exceptions). Otherwise you don't have any control over what happens after exception being thrown and app may simply "explode".

jacek15157 replied:
Moreover I'm losing the original stack trace of the exception, as well as its type here. I fixed it in a commit below. Thx for review.

EnjoyTheNoise reviewed (changes requested):
Uploading photos is not addressed here. You're expecting URL, but from where? Frontend? I guess no.
Other thing is, by deleting existing endpoint for checking email, you made sagas unuseable without changes, not to say, there's none method that can replace old one.

EnjoyTheNoise reviewed (changes requested):
Also what I found on frontend: you're sending POST to image-upload endpoint, which does not exist at all in this PR, care to explain?

jacek15157 commented:
> Uploading photos is not addressed here. You're expecting URL, but from where? Frontend? I guess no.
> Other thing is, by deleting existing endpoint for checking email, you made sagas unuseable without changes, not to say, there's none method that can replace old one.

Still don't know how to handle uploading images, on front or back end, so it is a temporary solution. I need a talk about it. Endpoint for saga will come back in a minut.

jacek15157 committed:
Added CheckIfEmailExist, fixed some mistake in update user

EnjoyTheNoise committed:
Fixed namespace

EnjoyTheNoise committed:
Added UserDto for getting user from db

EnjoyTheNoise committed:
GetUser now returns dto, DeleteUser use email instead of full object

fixed namespace, optimised returned result in GetUser (redundant mapper call), generalized retrieving user from db

EnjoyTheNoise committed:
Removed unnecessary properties from update dto, fixed namespace

EnjoyTheNoise committed:
Update returns OK instead of Created, Delete uses email in whole flow

Added template matching for method attributes, fixed namespace

EnjoyTheNoise reviewed (approved):
Pushed my changes, imho it can be merged but leaving this for further review

jacek15157 commented:
> Pushed my changes, imho it can be merged but leaving this for further review

looks fine
